lib/spack/spack/test/llnl/util/link_tree.py

```diff
@@ -396,3 +397,257 @@ def test_source_merge_visitor_does_deals_with_dangling_symlinks(tmp_path: pathli
+        contents = os.listdir(head)
+        if tail not in contents:
+            return False
+    return True
+
+
+def test_is_folder_on_case_sensitive_filesystem(tmp_path: pathlib.Path, monkeypatch):
+    # we need to mock out both case sensitive and case insensitive cases since
+    # we could be running on either
+
+    with monkeypatch.context() as m:
+        m.setattr(os.path, "exists", _mock_exists_case_insensitive)
+        assert is_folder_on_case_insensitive_filesystem(str(tmp_path))
+
+    with monkeypatch.context() as m:
+        m.setattr(os.path, "exists", _mock_exists_case_sensitive)
+        assert not is_folder_on_case_insensitive_filesystem(str(tmp_path))
+
+
+def test_source_visitor_no_path_normalization(tmp_path: pathlib.Path):
+    src = str(tmp_path / "a")
+
+    a = SourceMergeVisitor(normalize_paths=False)
+    a.visit_file(src, "file", 0)
+    a.visit_file(src, "FILE", 0)
+    assert len(a.files) == 2
+    assert len(a.directories) == 0
+    assert "file" in a.files and "FILE" in a.files
+    assert len(a.file_conflicts) == 0
+
+    a = SourceMergeVisitor(normalize_paths=False)
+    a.visit_file(src, "file", 0)
+    a.before_visit_dir(src, "FILE", 0)
+    assert len(a.files) == 1
+    assert "file" in a.files and "FILE" not in a.files
+    assert len(a.directories) == 1
+    assert "FILE" in a.directories
+    assert len(a.fatal_conflicts) == 0
+    assert len(a.file_conflicts) == 0
+
+    # without normalization, order doesn't matter
+    a = SourceMergeVisitor(normalize_paths=False)
+    a.before_visit_dir(src, "FILE", 0)
+    a.visit_file(src, "file", 0)
+    assert len(a.files) == 1
+    assert "file" in a.files and "FILE" not in a.files
+    assert len(a.directories) == 1
+    assert "FILE" in a.directories
+    assert len(a.fatal_conflicts) == 0
+    assert len(a.file_conflicts) == 0
+
+    a = SourceMergeVisitor(normalize_paths=False)
+    a.before_visit_dir(src, "FILE", 0)
+    a.before_visit_dir(src, "file", 0)
+    assert len(a.files) == 0
+    assert len(a.directories) == 2
+    assert "FILE" in a.directories and "file" in a.directories
+    assert len(a.fatal_conflicts) == 0
+    assert len(a.file_conflicts) == 0
+
+
+def test_source_visitor_path_normalization(tmp_path: pathlib.Path, monkeypatch):
+    src = str(tmp_path / "a")
+
+    # file conflict with os.path.samefile reporting it's NOT the same file
+    a = SourceMergeVisitor(normalize_paths=True)
+    a.visit_file(src, "file", 0)
+    a.visit_file(src, "FILE", 0)
+    assert a.files
+    assert len(a.files) == 1
+    # first file wins
+    assert "file" in a.files
+    # this is a conflict since the files are reported to be distinct
+    assert len(a.file_conflicts) == 1
+    assert "FILE" in [c.dst for c in a.file_conflicts]
+
+    # file conflict with os.path.samefile reporting it's the same file
+    a = SourceMergeVisitor(normalize_paths=True)
+    a.visit_file(src, "file", 0)
+    with monkeypatch.context() as m:
```

Comment:
I would just create the files then. `SourceMergeVisitor` is not symbol/abstract.

Reply:
Right but I need to test this code path in the visitor no?

The problem with using the actual file system is that the tests then depend on the file system that they run on, whereas with the direct calls to the visit function I can test this independently. 

I don't I can prepare two files with different capitalization but same inode without being on a case insensitive file system, that's why I'm using the money patch.

Reply:
use two directories ;)

Reply:
But the filesystem access will not be case insensitive in that case will it? That means the tests can only run on macOS, does it not?

Reply:
I think it should be OK, cause `os.path.samefile` is about inodes, not filenames. That's the only file system operation it's doing, right?

Reply:
I think I now understood what you were suggesting. I implemented the test without the mock now, and it works at least on macOS (but you're right it should also work on Linux I think).